src/trusted/service_runtime/nacl_app_thread.c

```diff
@@ -107,14 +108,30 @@ struct NaClApp *NaClChildNapCtor(struct NaClApp *nap, int child_cage_id, enum Na
 
   NaClXMutexUnlock(&nap_parent->children_mu);
 
-  NaClLog(1, "child_cage_id = %d, cage_id = %d\n", child_cage_id, nap_child->cage_id);
-  //Prevalidate forked nexe to remove loading time. We know it's valid because the parent is
-  nap_child->main_exe_prevalidated = 1;
-  if ((*mod_status = NaClAppLoadFileFromFilename(nap_child, nap_child->nacl_file)) != LOAD_OK) {
-    NaClLog(1, "Error while loading \"%s\": %s\n", nap_child->nacl_file, NaClErrorString(*mod_status));
-    NaClLog(LOG_FATAL, "%s\n%s\n",
-                       "Using the wrong type of nexe (nacl-x86-32 on an x86-64 or vice versa) ",
-                       "or a corrupt nexe file may be responsible for this error.");
+  NaClLog(1, "fork_num = %d, cage_id = %d\n", fork_num, nap_child->cage_id);
+  if(!use_lkm || tl_type != THREAD_LAUNCH_FORK) {
+    //exec not prevalidated
+    if ((*mod_status = NaClAppLoadFileFromFilename(nap_child, nap_child->nacl_file)) != LOAD_OK) {
```

Comment:
with the lkm we don't need to do the AppLoadFile part anymore?

Reply:
Yeah... not for fork, the reason why is that we don't need to load the file, we're just copying the program from the place it was in the parent cage, below this I set all the flags that NaClAppLoadFileFromFilename sets that are needed for this to work.